### PR TITLE
Patching clip model to create mask tensor on the device

### PR DESCRIPTION
**Pull request by shanmugamr1992**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by amyeroberts** [comment]
@shanmugamr1992 Thanks for opening this PR! 

For the quality tests to pass, you'll need to run `make fixup` locally and commit any formatting changes that were applied. There were some recent updates to our formatting libraries, so you might need to run `pip install -e .["quality"]` to ensure that you have the up-to-date settings. 

Could you add this update to the other implementations of `_build_causal_attention_mask` for pytorch models in the repo too please?

**Comment by shanmugamr1992** [comment]
The main idea is to reduce host and device syncs

**Comment by shanmugamr1992** [comment]
@amyeroberts  Could you verify and merge it please thanks .

File: src/transformers/models/clipseg/modeling_clipseg.py [hunk 1]
```diff
@@ -764,10 +764,10 @@ def forward(
             attentions=encoder_outputs.attentions,
         )
 
-    def _build_causal_attention_mask(self, bsz, seq_len, dtype):
+    def _build_causal_attention_mask(self, bsz, seq_len, dtype, device=None):
         # lazily create causal attention mask, with full attention between the vision tokens
         # pytorch uses additive attention mask; fill with -inf
-        mask = torch.empty(bsz, seq_len, seq_len, dtype=dtype)
+        mask = torch.empty(bsz, seq_len, seq_len, dtype=dtype, device=device)
         mask.fill_(torch.tensor(torch.finfo(dtype).min))
```
Comment (amyeroberts) [inline src/transformers/models/clipseg/modeling_clipseg.py:771]:
```suggestion
        mask.fill_(torch.finfo(dtype).min)
```

**Review by amyeroberts — APPROVED** [review]
LGTM! Thanks for adding this improvement and updating across the models. 

Two small comments:
* I've added a small suggestion to include your cleaner version of the mask fill for clipseg. 
* The diff in the README's is likely due to being slightly out-of-date with main - should be fixed with a quick rebase. 

Ready to merge once those two things are resolved :)

**Comment by amyeroberts** [comment]
@shanmugamr1992  Thanks again for your contribution and updating. I think just a final rebase / fix conflicts as there were some recent updates on the docs on `main`, then we're good to merge :)

**Comment by shanmugamr1992** [comment]
@amyeroberts  All done. Thanks a lot for the wonderful suggestions.

File: src/transformers/models/clip/modeling_clip.py [hunk 2]
```diff
@@ -714,8 +714,8 @@ def forward(
         bsz, seq_len = input_shape
         # CLIP's text model uses causal mask, prepare it here.
         # https://github.com/openai/CLIP/blob/cfcffb90e69f37bf2ff1e988237a0fbe41f33c04/clip/model.py#L324
-        causal_attention_mask = self._build_causal_attention_mask(bsz, seq_len, hidden_states.dtype).to(
-            hidden_states.device
+        causal_attention_mask = self._build_causal_attention_mask(
```
Comment (blefaudeux) [inline src/transformers/models/clip/modeling_clip.py:717]: cc @kashif, if the mask was bool here then the tower would be bf16 compatible I believe
